056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+This is causing a problem in combined mode because if the quorum has not formed due to some of the nodes not being ready 
+the KafkaRoller will still try to contact the broker via the admin client.
+This call fails because the quorum is not formed, so in some cases this results in the cluster being stuck with some nodes 
+in a pending state.
+
+## Motivation
+
+When running in the KRaft mode the controller pods need to be rolled if configuration changes occur, also at the moment the 
+existing logic is blocking the full implementation of liveness and readiness checks as described in [proposal #46](https://github.com/strimzi/proposals/blob/main/046-kraft-liveness-readiness.md).
+
+## Proposal
+
+The KafkaRoller behaviour should be unchanged when operating against a ZooKeeper based cluster.
+
+The behaviour when operating against a KRaft cluster is described below.
+
+### Order to roll nodes
+Roll in order:
+1. Unready controller/combined nodes
+2. Ready controller/combined nodes
+3. Active controller (applies to both pure controller and combined node)
+4. Unready broker-only nodes
+5. Ready broker-only nodes
+
+
+### Triggers
+The following are some of the triggers that would roll a KRaft controller or combined node:
+- Pod or its StrimziPodSet is annotated for manual rolling update
+- Pod is unready 
+- Controller's configuration has changed
+
+The triggers for broker remain the same as Zookeeper mode.
+
+### Rollability checks
+The checks made by the KafkaRoller in different modes is described below.
+
+#### The new quorum check
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Create admin client connection to the brokers and call `describeMetadataQuorum` API
+- If failed to connect to the brokers, return `UnforceableProblem` which will result in delay and retry for the pod until the maximum attempts is reached.
+- From the quorum info returned from the admin API, fetch `lastCaughtUpTimestamp` for every controller except the current controller we are considering to roll. `lastCaughtUpTimestamp` is the last millisecond timestamp at which a replica controller was known to be caught up with the quorum leader.
+- Check the quorum leader id using the quorum info and identify the `lastCaughtUpTimestamp` of the quorum leader
+- Retrieve value of the Kafka property `controller.quorum.fetch.timeout.ms` from the desired configurations specified in the Kafka CR. If this property does not exist in the desired configurations, then use the hard-coded default value for it which is `2000`. The reason for this is explained further in the **NOTE** below.
+- Mark a node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`.
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`).
+- Can roll if: `numOfCaughtUpControllers > ceil((double) totalNumOfControllers / 2)`.
```

Comment:
Is this correct? 
If totalNumOfControllers = 3, then `ceil((double) totalNumOfControllers / 2)` will be 2.
Then, we need 3 CaughtUpControllers to be able to roll?

Reply:
I think this is correct. If you roll with 2 controllers out of 3, the active controller will start a new election after quorum.fetch.timeout.ms has passed with no fetch requests from a majority of the quorum.

Reply:
The idea was actually to roll as long as there are 2 other controllers that are up to date with the leader if totalNumOfControllers = 3. So it's based on majority, rather than every node. It should be `numOfCaughtUpControllers.get() >= ceil((double) totalNumOfControllers / 2)`.

Reply:
I agree @tinaselenge, otherwise you won't roll at all, but you have to set `controller.quorum.fetch.timeout.ms` to something reasonably high while rolling. For example, this would avoid an unnecessary election when you have 2 caught up controllers, and 1 controller is down in a 3 controllers setup.

Reply:
`Maximum time without a successful fetch from the current leader before becoming a candidate and triggering an election for voters; Maximum time without receiving fetch from a majority of the quorum before asking around to see if there's a new epoch for leader`

This does not seem to say the quorum leader would trigger an election if it hasn't received fetch requests from the majority. What's majority for it in 3 controllers case though? Does it consider itself part of the majority? Only the followers could trigger an election if it can't successfully fetch so as long as the remaining follower is happily fetching from the leader, it shouldn't trigger an election?

Reply:
>  Does it consider itself part of the majority? 

Yes, it is.

> Only the followers could trigger an election if it can't successfully fetch so as long as the remaining follower is happily fetching from the leader

I think the active controller (KRaft leader) will step down and trigger a new election instead. This also described in the classic Raft algorithm to avoid zombie leaders in case of network partitions.

In KIP-595 we have the so called "progress timeout". When a majority of voters are not fetching from the leader, the current leader will start election to reach out to other followers and re-establish the leadership or learn about the new leader. This is also the cause of an known issue described in the draft KIP-650 (that can be solved with pre-vote).

Reply:
I see. Thanks for clarifying that. 

So regarding your earlier point, are you saying when 1 controller is down and 1 remaining follower is fetching happily from the leader, the leader is receiving fetch requests from the remaining follower, that's considered majority of voters not fetching from the leader and would trigger re-election?

Reply:
> when 1 controller is down and 1 remaining follower is fetching happily from the leader, the leader is receiving fetch requests from the remaining follower, that's considered majority of voters not fetching from the leader and would trigger re-election

No. When you have 2 out of 3 controllers that are in sync, that's ok (quorum is healthy). That said, when you want to roll the cluster for whatever reason, you would start by restarting the remaining follower (leader last policy). If this node restart lasts for more than `controller.quorum.fetch.timeout.ms` (2 seconds by default), then the leader will start a new election.

If you increase `controller.quorum.fetch.timeout.ms` to include the node clean restart period, then no election should be triggered during a rolling update as the follower/leader will come back before the other node realizes it was alone. Unfortunately, this also means that you would increase leader failure/partition detection. Given that this timeout can't be changed dynamically, we would need to find a reasonable value, which may be tricky. Wdyt?

cc @showuon @tombentley @mimaison

Reply:
I think I understand it. Just to check if the proposal is describing the intended check correctly, I would like to re-iterate the logic here. We would not roll the follower that has caught up to the leader while another one is not in sync yet. For example, if we have controller-1 (leader), controller-2 (follower in sync), controller-3 (follower is not in sync), we would only allow controller-3 to be rolled, since quorum would still stay healthy while it's down. When attempting to roll controller-1 or controller-2, the quorum check will return false because this would leave the quorum without any in sync follower. We will block rolling until controller-3 has caught up to the leader.

Also `controller.quorum.fetch.timeout.ms` is a configuration we allow users to change today so would it make sense for Strimzi to set this configuration?

Reply:
> We would not roll the follower that has caught up to the leader while another one is not in sync yet.

Yes, that would avoid the new election in case `controller.quorum.fetch.timeout.ms` is exceeded. 

That said, in your example, what would happen if the leader fails while you are restarting a follower? The remaining follower would start an election, but won't be able to be elected as the new leader because 2 out of 3 nodes are down. This would be a service downtime. The new leader would be elected as soon as the controller you are restarting comes up again, restoring the service.

If you want to completely avoid any downtime, you need to check that a majority of controllers is in-sync (3/3, 3/5, 4/7, and so on).

> Also controller.quorum.fetch.timeout.ms is a configuration we allow users to change today so would it make sense for Strimzi to set this configuration?

Probably not, as we want to give users the flexibility to tune it, but only in very specific use cases (most users won't need to touch it).

Reply:
I think this is addressed now.